Remora.Rest.Core/Optional.cs

```diff
@@ -155,4 +183,86 @@ public override string ToString()
             ? $"{{{_value?.ToString() ?? "null"}}}"
             : "Empty";
     }
+
+    /// <summary>
+    /// Returns the value of the current <see cref="Optional{TValue}"/>, or <c>default</c> if the current
+    /// <see cref="Optional{TValue}"/> is empty.
+    /// </summary>
+    /// <returns>The value of <c>this</c> or <c>default</c> if none is present.</returns>
+    public TValue? OrDefault()
+    {
+        return TryGet(out var value) ? value : default;
+    }
+
+    /// <summary>
+    /// Returns the value of the current <see cref="Optional{TValue}"/>, or <paramref name="defaultValue"/> if
+    /// <see cref="Optional{TValue}"/> is <b>null or empty</b>.
+    /// </summary>
+    /// <param name="defaultValue">The default value to fallback to if the optional is empty.</param>
+    /// <returns>
+    /// The value of <see cref="Optional{TValue}"/> or <paramref name="defaultValue"/> if none is present.
+    /// </returns>
+    [return: NotNullIfNotNull(nameof(defaultValue))]
+    public TValue? OrDefault(TValue? defaultValue)
+    {
+        return IsDefined(out var value) ? value : defaultValue;
+    }
+
+    /// <summary>
+    /// Returns the value of the current <see cref="Optional{TValue}"/>, or throws the exception in
+    /// <paramref name="func"/> if empty.
+    /// </summary>
+    /// <param name="func">The function generating an <see cref="Exception"/>.</param>
+    /// <returns>The value of <see cref="Optional{TValue}"/>.</returns>
+    /// <exception cref="Exception">If <see cref="Optional{TValue}"/> is empty.</exception>
+    public TValue OrThrow([RequireStaticDelegate] Func<Exception> func)
+    {
+        return TryGet(out var value) ? value : throw func();
+    }
+
+    /// <summary>
+    /// Casts the current <see cref="Optional{TValue}"/> to a nullable <typeparamref name="TValue"/>?.
+    /// </summary>
+    /// <returns>
+    /// An <see cref="Optional{TValue}"/> with the type pararameter changed to <typeparamref name="TValue"/>?.
+    /// </returns>
+    public Optional<TValue?> Nullable()
```

Comment:
The workings of this method isn't entirely obvious - at first glance, I assumed it decayed an `Optional<T>` into a `Nullable<T>`. Some type of `Cast<T>` method might be more appropriate and flexible. I'd suggest `Map<T>`, too, but you've beat me to that one :P

Reply:
`Cast<T>` is not possible due because as far as i can tell it'd need a `where TValue : TResult` constraint which is impossible

Reply:
An unconstrained `Cast` would work (not every cast has a true polymorphic relationship), but then we're just back to `Map<T>`... I think a better name is just what we'll have to settle on.

`AsValueNullable`, perhaps? What's a real usecase for this method, just for context?

Reply:
> What's a real usecase for this method, just for context?

Providing an Optional<T> value when an Optional<T?> parameter is desired, for instance; in my code personally for instance, I have to turn `Optional<IAllowedMentions>` to `Optional<IAllowedMentions?>`, not doing so (as well as manually casting) will produce a nullability warning:
![image](https://user-images.githubusercontent.com/13633343/209485543-025a3bf5-724e-4e26-93c3-427c10d832d7.png)

`AsNullableOptional` might be a more clear name.